### PR TITLE
IPMPROG-4829  - fty shm rework

### PR DESCRIPTION
**Pull request by sinanyvs**

### CODE REVIEW

**Review by FrancoisRegisDegott-eaton — DISMISSED** [review]

File: benchmark/src/benchmark.cc [hunk 1]
```diff
@@ -163,11 +179,12 @@ void Benchmark::cpp_api_bench()
...
     for (i = 0; i < NUM_METRICS; i++) {
         char buf[METRIC_LEN];
-        sprintf(buf, METRIC_FMT, i);
+        snprintf(buf, sizeof(buf), METRIC_FMT, i);
         names.push_back(buf);
-        sprintf(buf, VALUE_FMT, i);
+        snprintf(buf, sizeof(buf), VALUE_FMT, i);
         values.push_back(buf);
     }
+    while(1);
```
Comment (ericclappier-eaton) [inline benchmark/src/benchmark.cc]: To remove  ?

File: lib/public_include/fty_shm.h [hunk 2]
```diff
@@ -22,64 +22,84 @@
...
+// @param metric : metric name (non NULL/empty)
+// @param value : value (non NULL/empty)
+// @param unit : unit (non NULL)
+// @param ttl : time to live (>=0)
+// Returns 0 on success, else -1 (sets errno accordingly)
 int fty_shm_write_metric(const char* asset, const char* metric, const char* value, const char* unit, int ttl);
 
+// Stores a single metric in shared memory, using the fty_proto_t model.
+// Same specifciations as fty_shm_write_metric()
+// @param proto : fty_proto_t object (non NULL)
```
Comment (ericclappier-eaton) [inline lib/public_include/fty_shm.h]: metric

File: lib/public_include/fty_shm.h [hunk 3]
```diff
@@ -22,64 +22,84 @@
...
+// Returns : 0 on success, -1 if metric/asset doesn't exist
 int fty_shm_read_metric(const char* asset, const char* metric, char** value, char** unit);
 
-// Use a custom storage directory for test purposes (the passed string must
-// not be freed)
+// Only UT or Debug
+// Use a custom storage directory for test purposes.
+// Should be called only on unit test.
+// Sets the current directory path
+// @param dir : the dirctory (non NULL, valid)
```
Comment (ericclappier-eaton) [inline lib/public_include/fty_shm.h]: directory

File: lib/src/fty_shm.cc [hunk 4]
```diff
@@ -248,84 +207,100 @@ int read_data_metric(const char* filename, fty_proto_t* proto_metric)
...
+    char filename[PATH_MAX];
+    if (build_metric_filename(filename, sizeof filename, asset, metric, FTY_SHM_METRIC_TYPE) < 0) {
         return -1;
-    if (!unit) {
-        char* dummy;
-        return read_value(filename, *value, dummy, false);
     }
-    return read_value(filename, *value, *unit);
+
+    fty_proto_t * proto_metric = fty_proto_new(FTY_PROTO_METRIC);
```
Comment (ericclappier-eaton) [inline lib/src/fty_shm.cc:266]: Add test for proto_metric allocation

File: lib/src/fty_shm.cc [hunk 5]
```diff
@@ -336,191 +311,249 @@ int fty_shm_read_family(const char* family, std::string asset, std::string type,
...
+        value = value_s ? value_s : "";
+    }
 
-    if (prepare_filename(
-            filename, asset.c_str(), asset.length(), metric.c_str(), metric.length(), FTY_SHM_METRIC_TYPE) < 0)
-        return -1;
-    return read_value(filename, value, dummy, false);
+    // cleanup
+    if (value_s) {
+        free(value_s);
```
Comment (ericclappier-eaton) [inline lib/src/fty_shm.cc:499]: delete ?

**Review by ericclappier-eaton — CHANGES_REQUESTED** [review]

File: lib/public_include/fty_shm.h (hunk 3 above)
Comment (sinanyvs) [reply lib/public_include/fty_shm.h]: didn't understand what you mean

File: lib/src/fty_shm.cc (hunk 4 above)
Comment (sinanyvs) [reply lib/src/fty_shm.cc:266]: read_data_metric applies a null check

File: lib/src/fty_shm.cc (hunk 5 above)
Comment (sinanyvs) [reply lib/src/fty_shm.cc:499]: didn't understand what you mean

Comment (ericclappier-eaton) [reply lib/src/fty_shm.cc:499]: Sorry it is ok, never see what you use strdup for allocation (malloc)

File: lib/public_include/fty_shm.h (hunk 3 above)
Comment (ericclappier-eaton) [reply lib/public_include/fty_shm.h]: typo error in the comment

**Review by ericclappier-eaton — APPROVED** [review]

**Review by aquette — APPROVED** [review]

**Review by FrancoisRegisDegott-eaton — APPROVED** [review]

**Pull request merged**